cdap-app-fabric/src/main/java/co/cask/cdap/app/services/AbstractServiceDiscoverer.java

```diff
@@ -85,4 +101,85 @@ private URL createURL(@Nullable Discoverable discoverable, String applicationId,
       return null;
     }
   }
+
+  private URL getStreamURL(String stream) throws StreamWriteException {
+    return getStreamURL(stream, false);
+  }
+
+  private URL getStreamURL(String stream, boolean batch) throws StreamWriteException {
+    ServiceDiscovered discovered = getDiscoveryServiceClient().discover(Constants.Service.STREAMS);
+    Discoverable discoverable = new RandomEndpointStrategy(discovered).pick(1, TimeUnit.SECONDS);
+    if (discoverable != null) {
+      InetSocketAddress address = discoverable.getSocketAddress();
+      String path = String.format("http://%s:%d%s/namespaces/%s/streams/%s", address.getHostName(), address.getPort(),
+                                  Constants.Gateway.API_VERSION_3, namespaceId, stream);
+      if (batch) {
+        path = String.format("%s/batch", path);
+      }
+      try {
+        return new URL(path);
+      } catch (MalformedURLException e) {
+        LOG.error("Got exception while creating StreamURL", e);
+      }
+    }
+    throw new StreamWriteException("Stream Service Endpoint not found");
+  }
+
+  private void writeToStream(String stream, HttpRequest request) throws IOException, StreamWriteException {
+    HttpResponse response = HttpRequests.execute(request);
+    if (response.getResponseCode() == HttpResponseStatus.NOT_FOUND.code()) {
```

Comment:
Can this ever happen? Don't you already check for stream existence when getting the URL?

Also, maybe throw exception on all writes that aren't 200s? Otherwise you can have failing writes without knowing.